bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/internal/discovery/SensorDiscoveryService.java

```diff
@@ -0,0 +1,139 @@
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.ecobee.internal.discovery;
+
+import static org.openhab.binding.ecobee.internal.EcobeeBindingConstants.*;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerService;
+import org.openhab.binding.ecobee.internal.dto.thermostat.RemoteSensorDTO;
+import org.openhab.binding.ecobee.internal.handler.EcobeeThermostatBridgeHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link SensorDiscoveryService} is responsible for discovering the Ecobee
+ * sensors that are assigned to a thermostat.
+ *
+ * @author Mark Hilbush - Initial contribution
+ */
+@NonNullByDefault
+public class SensorDiscoveryService extends AbstractDiscoveryService implements DiscoveryService, ThingHandlerService {
+
+    private final Logger logger = LoggerFactory.getLogger(SensorDiscoveryService.class);
+
+    private @Nullable EcobeeThermostatBridgeHandler bridgeHandler;
+
+    public SensorDiscoveryService() {
+        super(30);
+    }
+
+    @Override
+    public void setThingHandler(@Nullable ThingHandler handler) {
+        if (handler instanceof EcobeeThermostatBridgeHandler) {
+            ((EcobeeThermostatBridgeHandler) handler).setDiscoveryService(this);
+            bridgeHandler = (EcobeeThermostatBridgeHandler) handler;
+        }
+    }
+
+    @Override
+    public @Nullable ThingHandler getThingHandler() {
+        return bridgeHandler;
+    }
+
+    @Override
+    public void activate() {
+        logger.debug("SensorDiscovery: Activating Ecobee sensor discovery service");
+    }
+
+    @Override
+    public void deactivate() {
+        logger.debug("SensorDiscovery: Deactivating Ecobee sensor discovery service");
+    }
+
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypes() {
+        return SUPPORTED_SENSOR_THING_TYPES_UIDS;
+    }
+
+    @Override
+    public void startBackgroundDiscovery() {
+        EcobeeThermostatBridgeHandler localBridgeHandler = bridgeHandler;
+        if (localBridgeHandler == null) {
+            logger.info("SensorDiscovery: Can't perform background discovery because bridgeHandler is null");
```

Comment:
This is a good question.

First, it appears that I've implemented the `SensorDiscoveryService` differently than the `ThermostatDiscoveryService`. 😟  In one case, I used `@NonNullByDefault({})` in the `bridgeHandler` declaration. in the other case I made it `@Nullable`.

I'm a little unclear about something. In the old way of starting the discovery service, the discovery service was activated at the same time that the handler was created. That's because the handler factory did the handler creation AND the discovery service (e.g. see the squeezebox binding). 

Therefore, the discovery service could be certain that the handler was there when `startBackgroundDiscovery` was called. It's not clear to me if that's still the case with this new method of activating the discovery service. But isn't that the point of having the discovery service implement the `ThingHandlerService` interface. Do you know?

Reply:
Yea, there isn't any reasonable way that this would get called, might be appropriate to throw an InvalidStateException instead. I also think using `@NonNullByDefault({})` on the bridge handler field would be an appropriate solution as well.

Reply:
Done. Used `@NonNullByDefault({})`.